File: js/task3c/S3_PIPE_06.js
// S3_PIPE_06 - Safe pipeline
// catches errors and returns { ok, value/error }

const pipeSafe = (...fns) => x => {
  let result = x;

  for (const fn of fns) {
    try {
      result = fn(result);
    } catch (err) {
      return { ok: false, error: err.message };
    }
  }

  return { ok: true, value: result };
};

// test functions
const double = x => x * 2;
const addOne = x => x + 1;
const mustBePositive = x => {
  if (x <= 0) throw new Error("Value must be positive");
  return x;
};
const parseJSON = s => JSON.parse(s);

// success case
const safeMath = pipeSafe(double, addOne, mustBePositive);
console.log("safeMath(5):", safeMath(5));  // { ok: true, value: 11 }

// failure case
console.log("safeMath(-5):", safeMath(-5)); // { ok: false, error: "Value must be positive" }

// json parsing
const safeParseAndDouble = pipeSafe(parseJSON, x => x.value, double);
console.log('safeParseAndDouble(\'{"value": 10}\'):', safeParseAndDouble('{"value": 10}'));
console.log('safeParseAndDouble("bad json"):', safeParseAndDouble("bad json"));

File: python/task3c/S3_PIPE_06.py
# S3_PIPE_06 - Safe pipeline (pipe_safe)

def pipe_safe(*fns):
    def run(x):
        result = x
        for fn in fns:
            try:
                result = fn(result)
            except Exception as e:
                return {"ok": False, "error": str(e)}
        return {"ok": True, "value": result}
    return run

# Test functions
double = lambda x: x * 2
add_one = lambda x: x + 1

def must_be_positive(x):
    if x <= 0:
        raise ValueError("Value must be positive")
    return x

# Success case
safe_math = pipe_safe(double, add_one, must_be_positive)
print(f"safe_math(5): {safe_math(5)}")

# Failure case - raises exception
print(f"safe_math(-5): {safe_math(-5)}")

# Another failure case
def parse_int(s):
    return int(s)

safe_parse = pipe_safe(parse_int, double)
print(f"\nsafe_parse('10'): {safe_parse('10')}")
print(f"safe_parse('abc'): {safe_parse('abc')}")
